enhancements/baremetal/baremetal-poison-pill-remediation.md

```diff
@@ -0,0 +1,189 @@
+* To utilize the MHC logic for deciding if a node is unhealthy and whether remediation is appropriate
+* To allow the healthy majority to recover workloads after a pre-determined and finite interval
+* To allow a node to take actions to remediate itself in case of a failure
+* To allow a node to detect its health status when master<->node communication is partially failed
+* To avoid false positives caused by failure of the control plane
+
+### Non-Goals
+
+* Self healing in combination with existing reprovisioning or reboot based remediation mechanisms in the Machine Health Check controller
+* Recovering from all types of failure
+* Having a distributed mesh of health checks where each node is checking health for its peers
+* Creation of fake Machine objects for UPI clusters
+
+## Proposal
+
+Utilizing a DaemonSet to have a privileged critical pod running on each node. This pod will periodically check the
+health status of the node it’s running on, and tickle a [watchdog device](https://en.wikipedia.org/wiki/Watchdog_timer).
+
+This will be performed by looking for the `host.metal3.io/external-remediation` annotation added by MHC to the unhealthy
+Machine CR.
+
+If the unhealthy annotation was found, the node will remediate itself by using the watchdog device to
+trigger a reset of the node. Watchdog devices, especially the hardware kind, are considered more reliable and predictable than the
+reboot command.
+
+If Machine CR is not accessible in 3 consecutive attempts, it needs external assistance to detect its
+health status.
+
+Failure to read Machine CR could stem from one of the following:
+1. The node is unhealthy and therefore can’t read the Machine CR (e.g. local network failure)
+1. There’s an etcd failure
+1. Connectivity issues between the node and the control plane nodes (which doesn’t stem from local problem in the node)
+1. Resource starvation, preventing our DaemonSet from initiating the check 
+1. Authentication failure to API (ServiceAccount deleted?)
+1. RBAC disallows read operation on Machine (Role deleted or changed
+
+We only want to remediate if the problem is local (#1 and #4). Otherwise, we could create a "remediation storm" where
+all nodes are trying to remediate themselves, even if the problem is not theirs.
+
+The semantics of the watchdog device automatically handle #4 and to distinguish between the remaining three situations,
+the node will ask its peers whether they think it is healthy.
+
+The possible responses from a peer are:
+1. You're unhealthy (the annotation exists)
+1. You're healthy (the annotation does not exist)
+1. The peer also cannot contact the api-server
+1. No response (timeout or any TCP/HTTP error)
+
+To avoid saturating the network, the node will pick *min(cluster_size/10, cluster_size-1)* random nodes, and ask these
+nodes if they think it’s healthy using HTTP.
+Each pod will listen on a specific node-port, and will run an HTTP web server that will be able to tell other peers if
```

Comment:
Can we do it on multiple networks? Assuming (for example) there are multiple NICs, it might be that the 'external' network is unavailable, but the 'internal' is.

Reply:
Good point.
I guess it's possible and we should probably try all available IPs. 
I'll update the proposal accordingly